File: Core/ButtonCombo.swift
/// A combination of two or more buttons that must be held simultaneously
/// to trigger a composite action.
///
/// Combos are checked before single-button actions: if the held-button set
/// satisfies a registered combo, the combo action fires and the individual
/// button actions do not. This means single presses and combos never conflict
/// from the user's perspective.
///
/// Example:
/// ```swift
/// // A alone → new tab
/// // A + B together → bookmark (A's action is suppressed when B is pressed while A is held)
/// let bookmark = ButtonCombo(.a, .b)
/// ```
struct ButtonCombo: Hashable {

    /// The set of buttons that must all be held for this combo to fire.
    let buttons: Set<GamepadButton>

    /// Creates a combo from two or more buttons.
    init(_ first: GamepadButton, _ second: GamepadButton, _ rest: GamepadButton...) {
        self.buttons = Set([first, second] + rest)
    }

    // MARK: - Matching

    /// Returns `true` if every button in this combo appears in `heldButtons`.
    /// The held set may contain additional buttons beyond this combo's members.
    func matches(heldButtons: Set<GamepadButton>) -> Bool {
        buttons.isSubset(of: heldButtons)
    }
}

File: buttonbridge/core/button_combo.py
"""Multi-button combo type."""

from __future__ import annotations
from dataclasses import dataclass

from .gamepad_button import GamepadButton


@dataclass(frozen=True)
class ButtonCombo:
    """
    A set of buttons that must all be held simultaneously to trigger a
    composite action.

    Combos are checked before single-button actions: when a button press
    completes a registered combo, only the combo fires — the individual
    button actions do not.

    Example::

        # A alone → new tab.  A + B together → bookmark.
        combo = ButtonCombo(GamepadButton.A, GamepadButton.B)
    """

    buttons: frozenset[GamepadButton]

    def __init__(self, *buttons: GamepadButton) -> None:
        if len(buttons) < 2:
            raise ValueError("A combo requires at least two buttons.")
        object.__setattr__(self, "buttons", frozenset(buttons))

    def matches(self, held: set[GamepadButton]) -> bool:
        """Return True if every button in this combo is currently held."""
        return self.buttons.issubset(held)
